### DISCUSSION TITLE
Winter CMS 1.2 is now available for general testing

### DISCUSSION THREAD

**Discussion by bennothommo** [wintercms/winter#558] [Announcements]
![1 2](https://user-images.githubusercontent.com/15900351/168710794-5a6963e3-bc5d-41a0-bcfd-edca30cb3eb1.png)

The 1.2 branch of Winter is now available for general testing! Built on top of Laravel 9, it supports PHP 8.0 and 8.1, and includes better support for Tailwind, Font Awesome 6, command autocompletion and a whole heap of performance tweaks and code cleanup, making this the most streamlined release of Winter yet!

Check out the [blog announcement](https://wintercms.com/blog/post/v12-laravel-9-support-now-available) and have a read of the [upgrade guide](https://github.com/wintercms/meta/blob/master/l9-upgrade-notes.md)!

**Comment by fosterushka** [comment]
Huzzah 🎉🥳

**Comment by Artful-dodger** [comment]
Just reading the blurb on this release and was wondering how you scaffold a tailwindcss project as one of the bullet point suggests. Is there a spercific command for this?

**Reply by bennothommo** [reply]
Run the following: `php artisan create:theme [theme name] tailwind`

**Reply by Artful-dodger** [reply]
Thank you Ben for your prompt repsonse but I have tried that but get "Too many arguments to "create:theme" command, expected arguments "theme"."

**Reply by Artful-dodger** [reply]
I thought there maybe some filter commad like -- but nothing I have tried works

**Reply by bennothommo** [reply]
@Artful-dodger does `php artisan create:theme mytheme tailwind` work?

**Reply by Artful-dodger** [reply]
No it gives me the same response "Too many arguments to "create:theme" command, expected arguments "theme"
one thing I have notice it give this reponse for verion number
*** Detecting Winter CMS build...
*** We could not accurately determine your Winter CMS build due to the number of modifications. The closest detected build is Winter CMS build 1.1.8.
is this correct I have follwed the upgrade guide

**Reply by bennothommo** [reply]
Yeah it looks like you're running 1.2, but when did you do the upgrade?

**Reply by Artful-dodger** [reply]
I have manage to get it working from the installer now struggling with getting the mix to work I think this is a windows issue I'll try and figure this out
  file_put_contents(C:\wamp64\www\winter1-2\C:\wamp64\www\winter1-2\themes\test/mix.webpack.js): Failed to open stream: No such file or directory

**Reply by Artful-dodger** [reply]
looks like this is not compatible with wamp because pcntl extension is required and is not compatible with windows, I discovered this when trying php artisan mix:watch theme-tailwind gives an error of Undefined constant "System\Console\SIGINT"

**Reply by LukeTowers** [reply]
Can you submit an issue for that @Artful-dodger? 

Reading up on https://stackoverflow.com/a/37464124, https://www.reddit.com/r/PHP/comments/aavqes/its_impossible_in_php_to_handle_ctrlc_on_windows/, https://bugs.php.net/bug.php?id=77365, & https://www.php.net/manual/en/function.sapi-windows-set-ctrl-handler.php I think our best option is to implement the script termination logic in as cross platform as a way as we can in the base `Winter\Storm\Console\Command` class (or via a trait) and call a method when defined on the command class being processed when a user termination signal is sent and then switch the watch command to use that instead. @bennothommo @jaxwilko any thoughts?

**Reply by Artful-dodger** [reply]
I have tried to run this command "php artisan create:theme mytheme tailwind" on ubuntu 22.04 server and get the following error "file does not exist at path /var/www/html/modules/cms/console/scaffold/theme/tailwind/readme.stub" but when I check the stub file does exist not sure what is causing this error but the command does not complete

**Reply by Artful-dodger** [reply]
from what I have notice the error is caused by the stub file name being in capital letters and being called in lower case letters ie the file is README.stub but the script is looking for readme.stub not sure if this is just affected in ubuntu 22/04

**Reply by LukeTowers** [reply]
@Artful-dodger the issue is fixed on the develop branch

**Comment by Artful-dodger** [comment]
This morning just going to try the Installer file I have just downloaded see if that has any effect

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Reply by LukeTowers** [reply]
The installer won't give you 1.2. Try running composer update and then check modules/cms/console/create theme.php and compare it to the 1.2 branch on github.

**Reply by Artful-dodger** [reply]
Ok will do

**Reply by Artful-dodger** [reply]
Yes thats working now to just have to sort out the mix to work on windows but thank you for your prompt response

**Reply by Artful-dodger** [reply]
I am getting this error "file_put_contents(C:\wamp64\www\winter1-2\C:\wamp64\www\winter1-2\themes\test/mix.webpack.js): Failed to open stream: No such file or directory" when I run this command "php artisan mix:compile -p theme-test --production" and I have search for this file but can only find mix.webpack.js.fixtures am I missing something or is there a problem with wamp

**Reply by Artful-dodger** [reply]
is the error being caused by the fact that it is lookin for a file in "C:\wamp64\www\winter1-2\C:\wamp64\www\winter1-2\"  when it should be looking in there "C:\wamp64\www\winter1-2\themes\test/mix.webpack.js" if so how would go about fixing this issue

**Reply by LukeTowers** [reply]
The problem is in the path generation logic within the Mix command classes, you would need to do some debugging on them to figure out why it's treating an absolute path (C:\wamp64\www\winter1-2\themes\test) as a relative path (since it's adding it to the application base path of C:\wamp64\www\winter1-2). It's an issue only for Windows users because of the difference in how paths are handled between Unix systems and Windows. Unfortunately, none of the core team develops with Windows machines actively so it's not easy for us to find and solve those issues so your best chance at getting it solved is to step through the code for the System\Console\MixCompileCommand class yourself and figure out where it's getting confused and then update us on your findings or submit a PR.

**Reply by Artful-dodger** [reply]
Thank you Luke I will do this over the week end

**Reply by Artful-dodger** [reply]
I had a bit of spare time and found where the error is accuring its line 114
   /**
     * Get the path to the mix.webpack.js file for the provided winter.mix.js file
     */
    protected function getWebpackJsPath(string $mixJsPath): string
    {
        return $this->getPackagePath($mixJsPath) . '/mix.webpack.js';
    }
not sure yet how to fix but this is where the duplicate path comes from

**Reply by LukeTowers** [reply]
@Artful-dodger if you could also submit this as a separate issue that would be helpful.

**Reply by Dreanad** [reply]
Hey there, its from the MixCompile.php file on line 94, the base_path() method add the duplication of C:\ 
because the base_path function doesnt exist, its returning the absolut path of the app ( helpers from laravel) 

i dont know how to resolve this properly, hope that can help  i will try on a linux distribution and my windows to find a solution

**Comment by Artful-dodger** [comment]
Hi Louis,

Not sure the cause of the problem as yet as I have installed Laravel-mix on wamp as a stand alone project and it works fine I am going to reinstall the tailwind theme again and see if I can get it to work with configuration I have setup within wintercms 1.2 let you know how I get on also there is an issue with the watch command not sure if they are both linked

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

________________________________

**Comment by ivang76** [comment]
Hi there,
I'm trying to update and test this new version on a running winter project, already migrated from october some months ago, but when I run the composer update I get this error:

```
In WebDAVAdapter.php line 18:
    Class "League\Flysystem\Adapter\AbstractAdapter" not found   

```
In the upgrade guide I read that "The Rackspace Flysystem driver / adapter is no longer supported", ok, but then what I should do to fix this error? I'm gonna to follow this one: https://laravel.com/docs/9.x/upgrade#flysystem-3

Thanks!

**Reply by ivang76** [reply]
ok I solved simply adding this:
`composer require league/flysystem-webdav:^3.0`

and then, depending by the needs:
```
Amazon S3: composer require -W league/flysystem-aws-s3-v3 "^3.0"
FTP: composer require league/flysystem-ftp "^3.0"
SFTP: composer require league/flysystem-sftp-v3 "^3.0"
```

**Comment by ivang76** [comment]
Now I have another problem... all the pages (created with Pages or with the CMS section) are going to 404 in both the two activated languages. 
There are no problems with the routes related to the backend.
Any ideas on what to check or debug?

If I disable the second language (but I would need at least two of them), it seems that something is working on the home page, but opening a page I get this error:
`Too few arguments to function Twig\Environment::loadTemplate(), 1 passed in /../.../testing_L9/plugins/winter/pages/classes/Controller.php on line 79 and at least 2 expected`

On the upgrade guide I read that "_The signature for $twig->loadTemplate() has been changed, use $twig->load() instead._" but should not be it already fixed in the plugin itself? (I already installed the latest version of twig)

**Reply by damsfx** [reply]
Do you require the 1.2 version for page plugin?  
`"winter/wn-pages-plugin": "dev-wip/1.2"`

**Reply by ivang76** [reply]
not via composer... I still use the plugin manager in the backend, where I see there is the version 2.0, is not anymore a working way?

**Reply by bennothommo** [reply]
Not until our own marketplace is live. It's probably good practice to get used to Composer though :)

**Reply by ivang76** [reply]
yes indeed... ok thanks, now that twig error is fixed, but if I re-add the second language, all the routes go in 404.
Maybe I have to update also the translate plugin via composer?

**Reply by bennothommo** [reply]
Yes, you will need to use `"winter/wn-translate-plugin": "dev-wip/1.2"` as well.

**Reply by LukeTowers** [reply]
@ivang76 Yes, see https://github.com/wintercms/wn-translate-plugin/commit/a1e8ef7107308ddb1b40087bf8885ba932b354db. Switch the translate plugin to the same `dev-wip/1.2`

**Reply by ivang76** [reply]
ok great, it works, thanks!

**Comment by ivang76** [comment]
I get again the abstract adapter error during the testing of winter 1.2, on another old plugin that I use from the old "october" times... it's the backup by Panagiotis [](url)https://github.com/panakour/oc-backup-plugin)
```
Class "League\Flysystem\Adapter\AbstractAdapter" not found
/Applications/MAMP/htdocs/Institutes_L9/plugins/panakour/backup/vendor/spatie/flysystem-dropbox/src/DropboxAdapter.php line 12
```
did maybe someone encountered and solved already this problem? or do you know if there is another similar plugin winter-laravel9 ready?

**Reply by LukeTowers** [reply]
If you can install panakour backup via composer instead and remove the vendor directory from the plugin itself then that will potentially solve your problem.

**Reply by ivang76** [reply]
I don't think it's installable via composer, and I see that it's a while now without updates or maintenance... 
so probably the best idea is to try to switch off that dropbox feature (that I'm not using) in the plugin sources, or maybe use the cool spatie package directly in background (but we will miss the ui)

**Reply by ivang76** [reply]
ok, as quick workaround if could be useful to someone else (and if you don't need dropbox), just change the line 11 of the file Repository.php (in the root of the panakour/backup) removing the dropbox function and rewrite it in this way:
```
    public function getAll()
    {
        return array_merge($this->getLocalBackups(), $this->getWebdavBackups());
    }
```

btw guys, is there some timeline about the final release (non beta) of the WinterCMS based on Laravel9? 
thanks!

**Reply by LukeTowers** [reply]
End of this month at the latest @ivang76, you can view the remaining items to be addressed before we merge it at the PR here: https://github.com/wintercms/winter/pull/148

FWIW the core maintainers and many community members are already using it in production.

**Reply by ivang76** [reply]
Hi @LukeTowers now that has been released, what's about the winter plugins we used during the test?
I mean, now in my composer I have them set as dev-wip. Can we update them to the latest "normal" version, for example these?
"winter/wn-pages-plugin": "^2.0",
"winter/wn-translate-plugin": "^2.1"

thanks!

**Reply by bennothommo** [reply]
@ivang76 we've released new versions of those plugins that support v1.2, so you can swap them back to versions now as opposed to the `dev` branches.

**Reply by ivang76** [reply]
Hi again @bennothommo and @LukeTowers , I would need your help again...

which is the best and correct way to upgrade those plugin, if I'm upgrading a system where there were the old rainlab version of them?
In particular for the "wm-pages-plugin" I simply installed via composer but I noticed that now my system miss the custom menu and all the pages created previously (that are still on the hard disk as files, but are not listed in the backend section).
I tried both keeping the old rainlab together then removing it and leaving only the new winter version...

**Reply by LukeTowers** [reply]
@ivang76 install the Winter plugins via composer and then just delete the RainLab plugin folders and run `artisan migrate` and it should work.

**Reply by damsfx** [reply]
@ivang76 to avoid having to think about it each time, you can modify your composer.json file to remove Rainlab's plugins when you run `composer update`
```json
        "post-update-cmd": [
            "rm -rf plugins/rainlab",
            "composer dump-autoload",
            "@php artisan winter:version",
            "@php artisan package:discover"
        ],
```

**Reply by LukeTowers** [reply]
Why would you do that? Just remove the rainlab plugins from your project once, shouldn't need to continually remove them.

**Reply by ivang76** [reply]
@LukeTowers I have done in that way, but I still have this "orphan" static pages no more linked to the web app, no listed in the backend and without my custom main menu.
(thanks damsfx, useful indeed... but in this moment my problem is to migrate the old versione of those plugins without lose data)

**Reply by LukeTowers** [reply]
@ivang76 sounds like something else is going wrong, the Winter.Pages plugin should just work with your existing RainLab data, there's no namespacing on the data itself that would prevent it from working. Can you send me your theme where the files are present but unable to be detected by Winter.Pages?

**Reply by damsfx** [reply]
> Why would you do that? Just remove the rainlab plugins from your project once, shouldn't need to continually remove them.

Just because some plugins got Rainlab User/Location/Page plugins in their package.json dependencies.  
Every time you run `composer u` they are reinstalled and then results in an issue if you run `php artisan winter:up` directly after.

**Reply by LukeTowers** [reply]
Gotcha. That should be fixed by adding replace definitions to the Winter plugin composer.json files.

**Reply by ivang76** [reply]
> @ivang76 sounds like something else is going wrong, the Winter.Pages plugin should just work with your existing RainLab data, there's no namespacing on the data itself that would prevent it from working. Can you send me your theme where the files are present but unable to be detected by Winter.Pages?

thanks, I'll try to investigate and debug deeper then... some references or files are deleted during the upgrading, maybe when I run the composer... I can't share the theme because contains private pages of my client sorry.

**Comment by Artful-dodger** [comment]
I not sure if there is an error in the setup for "php artisan create:theme [theme name] tailwind" in ubuntu 22.04 but the script fails at the readme.stub section because the file name is README.stub but the script is looking for readme.stub

**Reply by LukeTowers** [reply]
It's fixed on develop
